service/history/retry.go

```diff
@@ -117,7 +123,8 @@ func isRetryable(failure *failurepb.Failure, nonRetryableTypes []string) bool {
 }
 
 func getCauseFailure(failure *failurepb.Failure) *failurepb.Failure {
-	for ; failure.GetCause() != nil; failure = failure.GetCause() {
+	// Extract cause for ChildWorkflowExecutionFailure and ActivityFailure.
```

Comment:
Just to make sure I understand this - the intent here is to keep going until we hit the innermost childworkflowexecutionfailure or activityfailureinfo, right?

Reply:
No, it is vice versa. We moving "down" the chain while failure type is child execution of activity and there is an inner cause. As soon as we got something different from child execution or activity, we stop and return it. 
 
This is actual fix to the problem you reported. Before it goes down the last failure in the chain, which in your case, was generic go error converted to retryable `ApplicationFailure`.